File: app/infrastructure/telegram.py
"""Telegram notification adapter for sending video reports.

Sends annotated timelapse videos directly into a Telegram chat using
the ``python-telegram-bot`` library.  Gracefully no-ops when the bot
token or chat ID are not configured.  For videos exceeding 50 MB,
uploads to an S3-compatible storage provider and sends the public URL.
"""

from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

from loguru import logger

from app.core.config import settings

if TYPE_CHECKING:
    from app.infrastructure.storage import StorageProvider

try:
    import telegram

    _HAS_TELEGRAM = True
except ImportError:  # pragma: no cover
    _HAS_TELEGRAM = False


_UPLOAD_LIMIT = 50 * 1024 * 1024  # 50 MB — telegram.constants.FileSizeLimit.FILESIZE_UPLOAD


class TelegramNotifier:
    """Send messages and video files to a Telegram chat.

    Adapter over ``python-telegram-bot`` that encapsulates all Telegram
    wire calls so the rest of the application depends on this thin adapter
    rather than the third-party library directly.

    When ``settings.telegram_enabled`` is ``False``, or when the bot token /
    chat id are empty, every method is a no-op (logs at DEBUG level).

    When the video exceeds 50 MB, the notifier tries to upload it to an
    S3-compatible provider and sends the public URL as a text message.
    Falls back to the local download URL if no storage is configured.
    """

    def __init__(self, token: str, chat_id: str, storage: StorageProvider | None = None) -> None:
        self._token = token
        self._chat_id = chat_id
        self._storage = storage

    @classmethod
    def from_settings(cls) -> TelegramNotifier:
        """Build an instance from application settings.

        Also initialises a ``StorageProvider`` when storage is enabled
        so that oversize videos are uploaded automatically.

        Returns:
            A new ``TelegramNotifier`` wired to the current config.
        """
        storage: StorageProvider | None = None
        if settings.storage_enabled:
            from app.infrastructure.storage import StorageProvider as SP
            storage = SP.from_settings()
        return cls(settings.telegram_bot_token, settings.telegram_chat_id, storage=storage)

    def _enabled(self) -> bool:
        """Check whether Telegram notifications are configured and enabled."""
        return bool(
            settings.telegram_enabled
            and self._token
            and self._chat_id
            and _HAS_TELEGRAM
        )

    async def send_message(self, text: str) -> bool:
        """Send a plain text message to the configured chat.

        Args:
            text: The message body.

        Returns:
            ``True`` when the message was sent successfully, ``False``
            otherwise (or when the notifier is disabled).
        """
        if not self._enabled():
            logger.debug(f"Telegram disabled; would send message: {text[:80]}…")
            return False
        logger.info(f"Sending Telegram message: {text[:80]}…")
        try:
            bot = telegram.Bot(self._token)
            async with bot:
                await bot.send_message(chat_id=self._chat_id, text=text)
            return True
        except Exception:
            logger.exception("Failed to send Telegram message")
            return False

    async def send_video(
        self,
        path: Path,
        caption: str,
        fallback_url: str | None = None,
        public_url: str | None = None,
    ) -> bool:
        """Send a video file to the configured chat.

        If the file exceeds the 50 MB bot upload limit, tries to upload
        it to the configured S3-compatible storage provider and sends the
        public URL as a text message.  When storage is not configured or
        the upload fails, uses ``fallback_url`` as the download link.

        When the video is under the limit, the video is sent directly and
        a separate text message with the download URL is also sent so that
        the user always has a clickable link alongside the video preview.

        Args:
            path: Absolute filesystem path to the MP4 video.
            caption: Short description (max 1024 chars, shown under the
                video in the chat).
            fallback_url: Optional download URL for when the video exceeds
                the upload limit and no storage is configured.
            public_url: Optional public URL to send as a text message even
                when the video fits the upload limit.  Takes precedence
                over ``fallback_url``.

        Returns:
            ``True`` when the video (or a fallback message) was sent
            successfully, ``False`` otherwise.
        """
        if not self._enabled():
            logger.debug(f"Telegram disabled; would send video: {path}")
            return False

        size = path.stat().st_size if path.exists() else 0

        if size > _UPLOAD_LIMIT:
            logger.warning(
                f"Video {path} is {size / 1024 / 1024:.1f} MB, "
                f"exceeds 50 MB Telegram limit"
            )
            public_url_s3: str | None = public_url
            if not public_url_s3 and self._storage:
                logger.info("Attempting upload to S3-compatible storage")
                public_url_s3 = await self._storage.upload(path)

            link = public_url_s3 or fallback_url or caption
            msg = f"{caption}\n\n📹 Download: {link}"
            return await self.send_message(msg)

        logger.info(f"Sending video {path} ({size / 1024 / 1024:.1f} MB) to Telegram chat")
        try:
            bot = telegram.Bot(self._token)
            async with bot:
                with open(path, "rb") as fh:
                    await bot.send_video(
                        chat_id=self._chat_id,
                        video=fh,
                        caption=caption,
                        supports_streaming=True,
                        filename=path.name,
                        write_timeout=120,
                        read_timeout=120,
                    )
        except Exception:
            logger.exception(f"Failed to send video {path}")
            return False

        # Always send a text message with the download URL as well, so the
        # user has a clickable link even when the video was sent directly.
        link = public_url or fallback_url
        if link:
            msg = f"{caption}\n\n📹 Download: {link}"
            return await self.send_message(msg)

        return True

File: internal/infrastructure/telegram/telegram.go
// Package telegram sends video reports and plain messages to a Telegram
// chat via the Bot API, with a 50 MB upload cap and an S3 fallback for
// oversized files.
package telegram

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"time"

	"github.com/Lmex89/home-cameras/internal/config"
	"github.com/Lmex89/home-cameras/internal/infrastructure/storage"
)

// UploadLimit is Telegram's 50 MB bot upload cap.
const UploadLimit = 50 * 1024 * 1024

// Notifier sends messages and videos to a configured chat.
type Notifier struct {
	cfg     config.Config
	storage *storage.S3
	client  *http.Client
}

// New builds a notifier from the application config, wiring the S3
// provider when storage is enabled.
func New(cfg config.Config) *Notifier {
	var s3 *storage.S3
	if cfg.StorageEnabled {
		if p, err := storage.NewFromConfig(cfg); err == nil {
			s3 = p
		}
	}
	return &Notifier{
		cfg:     cfg,
		storage: s3,
		client:  &http.Client{Timeout: 120 * time.Second},
	}
}

// Enabled reports whether Telegram is configured and turned on.
func (n *Notifier) Enabled() bool {
	return n.cfg.TelegramEnabled && n.cfg.TelegramBotToken != "" && n.cfg.TelegramChatID != ""
}

func (n *Notifier) apiURL(method string) string {
	return fmt.Sprintf("https://api.telegram.org/bot%s/%s", n.cfg.TelegramBotToken, method)
}

// SendMessage posts a plain text message; returns false when disabled
// or on failure.
func (n *Notifier) SendMessage(ctx context.Context, text string) bool {
	if !n.Enabled() {
		return false
	}
	payload := map[string]any{"chat_id": n.cfg.TelegramChatID, "text": text}
	body, _ := json.Marshal(payload)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, n.apiURL("sendMessage"), bytes.NewReader(body))
	if err != nil {
		return false
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := n.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode >= 200 && resp.StatusCode < 300
}

// SendVideo sends an MP4 to the chat. Files over the 50 MB limit are
// uploaded to S3 when configured and reported as a link.
func (n *Notifier) SendVideo(ctx context.Context, path, caption, fallbackURL, publicURL string) bool {
	if !n.Enabled() {
		return false
	}
	size := int64(0)
	if info, err := os.Stat(path); err == nil {
		size = info.Size()
	}

	if size > UploadLimit {
		link := publicURL
		if link == "" && n.storage != nil {
			if url, err := n.storage.Upload(ctx, path); err == nil {
				link = url
			}
		}
		if link == "" {
			link = fallbackURL
		}
		return n.SendMessage(ctx, caption+"\n\nDownload: "+link)
	}

	if ok := n.sendVideoFile(ctx, path, caption); !ok {
		return false
	}
	link := publicURL
	if link == "" {
		link = fallbackURL
	}
	if link != "" {
		return n.SendMessage(ctx, caption+"\n\nDownload: "+link)
	}
	return true
}

func (n *Notifier) sendVideoFile(ctx context.Context, path, caption string) bool {
	f, err := os.Open(path)
	if err != nil {
		return false
	}
	defer f.Close()

	var buf bytes.Buffer
	mw := multipart.NewWriter(&buf)
	mw.WriteField("chat_id", n.cfg.TelegramChatID)
	mw.WriteField("caption", caption)
	mw.WriteField("supports_streaming", "true")
	part, err := mw.CreateFormFile("video", fileBase(path))
	if err != nil {
		return false
	}
	if _, err := io.Copy(part, f); err != nil {
		return false
	}
	mw.Close()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, n.apiURL("sendVideo"), &buf)
	if err != nil {
		return false
	}
	req.Header.Set("Content-Type", mw.FormDataContentType())
	resp, err := n.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode >= 200 && resp.StatusCode < 300
}

func fileBase(p string) string {
	for i := len(p) - 1; i >= 0; i-- {
		if p[i] == '/' || p[i] == '\\' {
			return p[i+1:]
		}
	}
	return p
}
